x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,287 @@
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+		}
+
+		// delete from the bonded validator index
+		store.Delete(GetBondedValidatorIndexKey(operator))
+
+		// update the validator
+		updates = append(updates, validator.ABCIValidatorUpdateZero())
+	}
+
+	return updates
+}
+
+// Validator state transitions
+
+func (k Keeper) bondedToUnbonding(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Bonded {
+		panic(fmt.Sprintf("bad state transition bondedToUnbonding, validator: %v\n", validator))
+	}
+	return k.beginUnbondingValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonding {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondedToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondedToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToUnbonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.completeUnbondingValidator(ctx, validator)
+}
+
+// send a validator to jail
+func (k Keeper) JailValidator(ctx sdk.Context, validator types.Validator) {
+	if validator.Jailed {
+		panic(fmt.Sprintf("cannot jail already jailed validator, validator: %v\n", validator))
+	}
+
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+	validator.Jailed = true
+	k.SetValidator(ctx, validator)
+	k.SetValidatorByPowerIndex(ctx, validator, pool)
+
+	// TODO we should be able to just delete the index, and only set it again once unjailed
+}
+
+// remove a validator from jail
+func (k Keeper) UnjailValidator(ctx sdk.Context, validator types.Validator) {
+	if !validator.Jailed {
+		panic(fmt.Sprintf("cannot unjail already unjailed validator, validator: %v\n", validator))
+	}
+
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+	validator.Jailed = false
+	k.SetValidator(ctx, validator)
+	k.SetValidatorByPowerIndex(ctx, validator, pool)
+}
+
+//________________________________________________________________________________________________
+
+// perform all the store operations for when a validator status becomes bonded
+func (k Keeper) bondValidator(ctx sdk.Context, validator types.Validator) types.Validator {
+
+	store := ctx.KVStore(k.storeKey)
+	pool := k.GetPool(ctx)
+
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+
+	validator.BondHeight = ctx.BlockHeight()
+
+	// XXX Are we OK with this? In order of power decreasing, but addresses break ties.
```

Comment:
What happened to the global counter?

Reply:
The global counter should probably be a separate PR - however, I've changed this to increment the intra-tx counter whenever a validator's power increases, which I think is a better solution than incrementing it in `bondValidator`.